### PR TITLE
error handling in kinsol

### PR DESCRIPTION
**Pull request by balos1**

### CODE REVIEW

File: src/kinsol/kinsol.c [hunk 1]
```diff
@@ -2562,44 +2569,36 @@ void KINPrintInfo(KINMem kin_mem, int info_code, const char* module,
...
+void KINProcessError(KINMem kin_mem, int error_code, int line, const char* func,
+                     const char* file, const char* msgfmt, ...)
 {
-  va_list ap;
-  char msg[256];
-
   /* Initialize the argument pointer variable
-     (msgfmt is the last required argument to KINProcessError) */
-
+     (msgfmt is the last required argument to cvProcessError) */
```
Comment (drreynolds) [inline src/kinsol/kinsol.c:2576]:
```suggestion
     (msgfmt is the last required argument to KINProcessError) */
```

**Review by drreynolds — CHANGES_REQUESTED** [review]

File: doc/kinsol/guide/source/Introduction.rst [hunk 2]
```diff
@@ -187,6 +187,10 @@ Users now need to link to ``sundials_core`` in addition to the libraries already
 This will be picked up automatically in projects that use the SUNDIALS CMake target. The library ``sundials_generic`` has been superceded by ``sundials_core`` and is no longer available.
 This fixes some duplicate symbol errors on Windows when linking to multiple SUNDIALS libraries.
 
+**Breaking change**
+The functions ``KINSetErrFile`` and ``KINSetHandlerErrFn`` have been removed.
+Users of these functions should use the :c:type:`SUNLogger` API instead.
```
Comment (gardner48) [inline doc/kinsol/guide/source/Introduction.rst:192]: Similar comment as in the ARKODE PR.

File: doc/kinsol/guide/source/Usage/index.rst [hunk 3]
```diff
@@ -527,57 +527,6 @@ negative, so a test ``retval`` :math:`<0` will catch any error.
   +--------------------------------------------------------+----------------------------------+------------------------------+
 
 
-.. c:function:: int KINSetErrFile(void * kin_mem, FILE * errfp)
```
Comment (gardner48) [inline doc/kinsol/guide/source/Usage/index.rst:530]: Looks like there are still some instances of `KINSetErrFile` and `KINSetErrHandlerFn` elsewhere in this file

File: src/kinsol/kinsol.c [hunk 4]
```diff
@@ -2562,75 +2558,33 @@ void KINPrintInfo(KINMem kin_mem, int info_code, const char* module,
...
-                   char* msg, void* data)
-{
-  KINMem kin_mem;
-  char err_type[10];
-
-  /* data points to kin_mem here */
-
-  kin_mem = (KINMem)data;
-
-  if (error_code == KIN_WARNING) { sprintf(err_type, "WARNING"); }
```
Comment (gardner48) [inline src/kinsol/kinsol.c:2624]: Unlike the other packages, I don't see any cases where KINSOL issues warnings.

File: src/kinsol/kinsol_ls.c [hunk 5]
```diff
@@ -1397,13 +1397,15 @@ int kinLs_AccessLMem(void* kinmem, const char* fname, KINMem* kin_mem,
 {
   if (kinmem == NULL)
   {
-    KINProcessError(NULL, KINLS_MEM_NULL, "KINLS", fname, MSG_LS_KINMEM_NULL);
```
Comment (gardner48) [inline src/kinsol/kinsol_ls.c:1400]: `fname` could be removed as an input to `kinLs_AccessLMem`

**Review by gardner48 — CHANGES_REQUESTED** [review]
Also noticed KINSOL does not have an error handler unit test like CVODE and ARKODE

**Review by drreynolds — APPROVED** [review]

File: src/kinsol/kinsol.c [hunk 6]
```diff
@@ -2571,16 +2575,31 @@ void KINProcessError(KINMem kin_mem, int error_code, int line, const char* func,
...
+  do {
+    if (kin_mem == NULL)
+    {
+      SUNGlobalFallbackErrHandler(line, func, file, msg, error_code);
+      break;
+    }
+
+    if (error_code == KIN_WARNING)
+    {
+#if SUNDIALS_LOGGING_LEVEL >= 1
```
Comment (gardner48) [inline src/kinsol/kinsol.c:2587]:
```suggestion
#if SUNDIALS_LOGGING_LEVEL >= SUNDIALS_LOGGING_WARNING
```

File: src/kinsol/kinsol_io.c [hunk 7]
```diff
@@ -875,7 +875,7 @@ int KINSetSysFunc(void* kinmem, KINSysFn func)
 
   if (func == NULL)
   {
-    KINProcessError(NULL, KIN_ILL_INPUT, __LINE__, __func__, __FILE__,
+    KINProcessError(kin_mem, KIN_ILL_INPUT, __LINE__, __func__, __FILE__,
```
Comment (gardner48) [inline src/kinsol/kinsol_io.c:878]: Good catch here and above!

**Review by gardner48 — CHANGES_REQUESTED** [review]

**Review by gardner48 — APPROVED** [review]

**Review by drreynolds — APPROVED** [review]

**Pull request merged**